Add Quirks for skype on iPadMini so the content is fully accessible

karlcow opened the pull request:
#### 390b5198a5960b5097fc77b379e3fe0d05a3752d
<pre>
Add Quirks for skype on iPadMini so the content is fully accessible
<a href="https://bugs.webkit.org/show_bug.cgi?id=275941">https://bugs.webkit.org/show_bug.cgi?id=275941</a>
<a href="https://rdar.apple.com/129048863">rdar://129048863</a>

Reviewed by Brent Fulgham.

Create a scrollable context for web.skype.com before starting a call,
so the bottom of the UI will be accessible for users.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsIPadSkypeOverflowScrollQuirk const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):

Canonical link: <a href="https://commits.webkit.org/281621@main">https://commits.webkit.org/281621@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/282d3837cf991299de561a1e3b60539c642a949c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60320 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39672 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12879 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64240 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10852 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62450 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47344 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11085 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48839 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7557 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62351 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36978 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52257 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29685 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33676 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9484 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9769 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55566 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9772 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65972 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4254 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9615 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56201 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4272 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52232 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56369 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13445 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3542 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35482 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36563 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37653 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36307 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/page/Quirks.h:
```diff
@@ -218,8 +219,10 @@ class Quirks {
 
     mutable std::optional<bool> m_hasBrokenEncryptedMediaAPISupportQuirk;
     mutable std::optional<bool> m_needsFullWidthHeightFullscreenStyleQuirk;
+    mutable std::optional<bool> m_needsIPadSkypeOverflowScrollQuirk;
```
I need to remove this once I'm able to test locally and uncomment below.

stwrt commented on Source/WebCore/style/StyleAdjuster.cpp:
```diff
@@ -928,6 +928,19 @@ void Adjuster::adjustForSiteSpecificQuirks(RenderStyle& style) const
+        // static MainThreadNeverDestroyed<const AtomString> selectorValue(".app-container .noFocusOutline > div"_s);
+        if (RefPtr div = dynamicDowncast<HTMLDivElement>(m_element)) {
+            auto matchesNoFocus = div->matches(".app-container .noFocusOutline > div"_s);
+            if (matchesNoFocus.hasException())
```
This function is gaining numerous separate quirks that require different behavior. Would it better to have separate functions for each website. I see this specific quirk function is handling Skype, Youtube, and WeChat.

brentfulgham replied:
This method could probably benefit from some refactoring, though not as part of this specific change. The domain-specific quirks (e.g., `needsGMailOverflowScrollQuirk`, `needsYouTubeOverflowScrollQuirk`, `needsWeChatScrollingQuirk`, etc.) seem like they would logically be mutually exclusive. Seems like a good idea for a follow-up patch, or something to consider with you and Karl's work on Quirks overall.

karlcow replied:
yes that's a good idea. I think there's a need for a basic function to override CSS in general.

stwrt reviewed (commented):
iPad mini should be two words in the commit message.

brentfulgham reviewed (approved):
Looks good. I think Brandon's suggestion that we refactor some of this code is good -- but we already were planning on that project to start soon! r=me

karlcow committed:
Add Quirks for skype on iPadMini so the content is fully accessible
https://bugs.webkit.org/show_bug.cgi?id=275941
rdar://129048863

Reviewed by Brent Fulgham.

Create a scrollable context for web.skype.com before starting a call,
so the bottom of the UI will be accessible for users.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsIPadSkypeOverflowScrollQuirk const):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjustForSiteSpecificQuirks const):

Canonical link: https://commits.webkit.org/281621@main
